storage/innobase/buf/buf0buf.cc

```diff
@@ -1878,39 +2088,99 @@ inline void buf_pool_t::resize()
+inline size_t buf_pool_t::garbage_collect()
 {
-  DBUG_ENTER("buf_resize_callback");
-  ut_ad(srv_shutdown_state < SRV_SHUTDOWN_CLEANUP);
-  mysql_mutex_lock(&buf_pool.mutex);
-  const auto size= srv_buf_pool_size;
-  const bool work= srv_buf_pool_old_size != size;
-  mysql_mutex_unlock(&buf_pool.mutex);
+  mysql_mutex_lock(&mutex);
+  lru_hp.set(nullptr);
+  size_t freed= 0;
 
-  if (work)
-    buf_pool.resize();
-  else
+  for (buf_page_t *bpage= UT_LIST_GET_LAST(LRU), *prev; bpage; bpage= prev)
   {
-    std::ostringstream sout;
-    sout << "Size did not change: old size = new size = " << size;
-    buf_resize_status(sout.str().c_str());
+    prev= UT_LIST_GET_PREV(LRU, bpage);
+    auto state= bpage->state();
+    ut_ad(state >= buf_page_t::FREED);
+    ut_ad(bpage->in_LRU_list);
+
+    /* We try to free any pages that can be freed without writing out
+    anything. */
+    switch (bpage->oldest_modification()) {
+    case 0:
+    try_to_evict:
+      if (buf_LRU_free_page(bpage, true))
+        freed++;
+      continue;
+    case 1:
+      break;
+    default:
+      if (state >= buf_page_t::UNFIXED)
+        continue;
+    }
+
+    if (state < buf_page_t::READ_FIX && bpage->lock.u_lock_try(true))
+    {
+      ut_ad(!bpage->is_io_fixed());
+      const lsn_t oldest_modification= bpage->oldest_modification();
+      switch (oldest_modification) {
+      case 1:
+        mysql_mutex_lock(&buf_pool.flush_list_mutex);
+        buf_pool.delete_from_flush_list(bpage);
+        mysql_mutex_unlock(&buf_pool.flush_list_mutex);
+        /* fall through */
+      case 0:
+        bpage->lock.u_unlock(true);
+        goto try_to_evict;
+      default:
+        if (bpage->state() < buf_page_t::UNFIXED &&
+            oldest_modification <= log_sys.get_flushed_lsn())
+          release_freed_page(bpage);
+        else
+          bpage->lock.u_unlock(true);
+      }
+    }
   }
-  DBUG_VOID_RETURN;
+
+#if defined MADV_FREE
+  /* FIXME: Issue fewer calls for larger contiguous blocks of
+  memory. For now, we assume that this is acceptable, because this
+  code should be executed rarely. */
+  for (buf_page_t *bpage= UT_LIST_GET_FIRST(free); bpage;
+       bpage= UT_LIST_GET_NEXT(list, bpage))
+    madvise(bpage->frame, srv_page_size, MADV_FREE);
+#endif
+  mysql_mutex_unlock(&mutex);
+  return freed;
+}
+
+/** Thread pool task invoked by innodb_buffer_pool_size changes. */
+static void buf_resize_callback(void *)
+{
+  buf_pool.resize();
 }
 
 /* Ensure that task does not run in parallel, by setting max_concurrency to 1 for the thread group */
 static tpool::task_group single_threaded_group(1);
 static tpool::waitable_task buf_resize_task(buf_resize_callback,
 	nullptr, &single_threaded_group);
 
+
 void buf_resize_start()
 {
-	srv_thread_pool->submit_task(&buf_resize_task);
+#if defined(UNIV_DEBUG) && defined(__linux__)
+  DBUG_EXECUTE_IF("trigger_gargbage_collection",
+  {
+    mem_pressure_obj.trigger_collection();
+  }
+  );
+#endif
```

Comment:
I would not add this. Let the resizing be entirely unchanged by this.

Reply:
I've still got the resize call back as a trigger for the debug test, but implementation is the same.